crates/ruff/src/rules/pyflakes/rules/unused_import.rs

```diff
@@ -192,52 +175,66 @@ pub(crate) fn unused_import(checker: &Checker, scope: &Scope, diagnostics: &mut
-    // creating a fix. We have to generate these diagnostics, even though they'll be ignored later
-    // on, so that the suppression comments themselves aren't marked as unnecessary.
-    for ((stmt_id, .., exceptions), unused_imports) in ignored
-        .into_iter()
-        .sorted_by_key(|((stmt_id, ..), ..)| *stmt_id)
+    // Separately, generate a diagnostic for every _ignored_ import, to ensure that the
+    // suppression comments aren't marked as unused.
+    for Import {
+        qualified_name,
+        trimmed_range,
+        parent_range,
+    } in ignored.into_values().flatten()
     {
-        let stmt = checker.semantic_model().stmts[stmt_id];
-        let multiple = unused_imports.len() > 1;
-        let in_except_handler =
-            exceptions.intersects(Exceptions::MODULE_NOT_FOUND_ERROR | Exceptions::IMPORT_ERROR);
-        for (qualified_name, range) in unused_imports {
-            let mut diagnostic = Diagnostic::new(
-                UnusedImport {
-                    name: qualified_name.to_string(),
-                    context: if in_except_handler {
-                        Some(UnusedImportContext::ExceptHandler)
-                    } else if in_init {
-                        Some(UnusedImportContext::Init)
-                    } else {
-                        None
-                    },
-                    multiple,
-                },
-                range,
-            );
-            if stmt.is_import_from_stmt() {
-                diagnostic.set_parent(stmt.start());
-            }
-            diagnostics.push(diagnostic);
+        let mut diagnostic = Diagnostic::new(
+            UnusedImport {
+                name: qualified_name.to_string(),
+                context: None,
+                multiple: false,
+            },
+            trimmed_range,
+        );
+        if let Some(range) = parent_range {
+            diagnostic.set_parent(range.start());
         }
+        diagnostics.push(diagnostic);
     }
 }
+
+/// An unused import with its surrounding context.
+struct Import<'a> {
+    /// The qualified name of the import.
+    qualified_name: &'a str,
+    /// The trimmed range of the import.
+    trimmed_range: TextRange,
+    /// The range of the import's parent statement.
+    parent_range: Option<TextRange>,
+}
+
+/// Generate a [`Fix`] to remove unused imports from a statement.
+fn fix_imports(checker: &Checker, stmt_id: NodeId, imports: &[Import]) -> Result<Fix> {
+    let stmt = checker.semantic_model().stmts[stmt_id];
+    let parent = checker.semantic_model().stmts.parent(stmt);
+    let edit = autofix::edits::remove_unused_imports(
+        imports
+            .iter()
+            .map(|Import { qualified_name, .. }| *qualified_name),
+        stmt,
+        parent,
+        checker.locator,
+        checker.indexer,
+        checker.stylist,
+    )?;
+    Ok(Fix::automatic(edit).isolate(checker.isolation(parent)))
+}
```

Comment:
I cleaned the implementation here up a bit, to use more of our modern conveniences (like `binding.trimmed_range`).